### PR TITLE
Add developmental test cases: idealized baroclinic wave and 2020 July CAPE cases + https://github.com/ufs-community/ufs-weather-model/pull/2459

### LINKED ISSUE #2309: Integration of UFS Case Studies ICs into UFS-WM
## Description
As a step toward an HTF/HSD framework for the UFS-WM, enabling the WM to run using GFS ICs from the [UFS Case Studies page](https://ufs-case-studies.readthedocs.io/en/develop/index.html) was identified as a desirable feature. This issue discusses a possible approach toward the development of this capability.

## Solution
To achieve this, ICs corresponding to the initialization date/time of a particular case need to be generated. One way to do this is with `chgres_cube`, which can be built as part of[ UFS_UTILS](https://github.com/ufs-community/UFS_UTILS). The conversion can be done automatically via the `gdas_init` utility, also available in UFS_UTILS. `gdas_init` can be configured to pull raw GFS data from HPSS onto a local machine (Hera or Jet) and convert the raw data (via `chgres_cube`) into the cold start format (including desired vertical level # and resolution) required by the ATM component of the UFS-WM. The converted cold start ICs can then be used as input data for certain UFS-WM RTs of corresponding model resolution and configuration, wherein the user need only replace the `sfc*.nc` and `gfs*.nc` files in the `INPUT` directory of the RT run directory and adjust the model start time in the `model_configure` file. Note that the develop version (as of 6/4/24) of UFS_UTILS generates the so-called `v2` `sfc` files via `gdas_init` and `chgres_cube`, so at this time, successful integration of the converted cold start files has only been achieved using the somewhat recently added `*v2.sfc` WM RTs (e.g., https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/tests/control_c48.v2.sfc; see #2005 and #1977). Since the `v2` surface files are significantly different from the `v1` surface file format, `v2` files do not seem to work with non-v2.sfc RTs.

To generate cold start ICs via UFS_UTILS/gdas_init/chgres_cube on an RDHPC with HPSS access (Hera or Jet), the user should do the following:

```
git clone https://github.com/ufs-community/UFS_UTILS.git
cd UFS_UTILS
cd fix
./link_fixdirs.sh emc <Hera or Jet>
cd ..
./build_all.sh
cd util
cd gdas_init
#edit the `config` file to set paths for data extraction and converted files (`EXTRACT_DIR` and `OUTDIR`, respectively); set  yy/mm/dd/hh to desired forecast start time +6 hours); set CDUMP to `gfs`  to generate GFS ICs; for the UFS-WM, set `LEVS` to 128; set `EXTRACT_DATA` to `yes`(unless data is already staged in `EXTRACT_DIR`); set `CRES_HIRES` to desired resolution for cold start files (e.g., C48)
# edit `PROJECT_CODE` in the machine-specific driver (e.g., `driver.jet.sh`) to an account user can use for submitting batch jobs; 
# run the driver script
./driver.<machine.sh
```

These steps will (1) pull raw GFS data from HPSS into `EXTRACT_DIR` and then (2) convert the raw data to cold start ICs (placed in `OUTDIR`, appended by `gfs.YYYYMMDD/HH/model_data/atmos/input`). To use the converted cold start files with the UFS-WM, copy the gfs*.nc and sfc*.nc files from the `OUTDIR/gfs.YYYYMMDD/HH/model_data/atmos/input` directory into the UFS-WM RT run directory's `INPUT` directory. Note that the resolution of the converted cold start GFS data must match the ATM resolution of the UFS-WM RT (e.g., use C48 GFS cold start data in the `control_c48.v2.sfc` RT). Then, edit the model start year/month/day/hour in the RT `model_configure`. The user can then re-run the `fv3.exe` via `sbatch job_card`. The case should now run at the new model start time with the cold start ICs. 

An example can be found in `/lfs4/HFIP/hfv3gfs/Cameron.Book/RT_RUNDIRS/Cameron.Book/FV3_RT/rt_1826477/control_c48.v2.sfc_intel`, which used cold start ICs generated through gdas_init/chgres_cube for the [2020 CAPE case](https://ufs-case-studies.readthedocs.io/en/develop/2020CAPE.html) (see /lfs4/HFIP/hfv3gfs/Cameron.Book/chgres_data/converted/gfs.20200723/00/model_data/atmos/input) from the UFS Case Studies page. 

Note that, in order to generate all necessary configuration/data/input/fix files to run in this method, the user first needs to run the base RT using `./rt.sh -n <test>'  to set-up the run directory (if working from ufs-wm/tests). The test cases will be integrated with the YAML-based testing feature as detailed in WM issue #2182 and PR #2278 (@jkbk2004).

If working in ufs-wm/tests-dev, the run directories will be setup when the user runs `ufs_test.sh -a <account> -s (if first time running) -n <test_name compiler> / -l <test YAML>`. `-n` should be used for a single test, while `-l` should be used if a user wishes to run from a YAML containing a list of tests.

## Related to
Issue #2182 and PR #2278

### PR DESCRIPTION
**Pull request by ulmononian**
## Commit Queue Requirements:

- [x] Fill out all sections of this template.
- [ ] All sub component pull requests have been reviewed by their code managers.
- [x] Run the full Intel+GNU RT suite (compared to current baselines) on either Hera/Derecho/Hercules
[RegressionTests_hera.log](https://github.com/user-attachments/files/17369313/RegressionTests_hera.1.log)

- [x] Commit 'test_changes.list' from previous step
---
## Description:
Add ATM-only idealized dry baroclinic wave test and a 2020 July CAPE case study from the [UFS Case Studies page](https://ufs-case-studies.readthedocs.io/en/develop/2020CAPE.html). The baroclinic wave is run at C192 resolution; 2020 July CAPE currently at C48 resolution using v2 surface files. The cases can be run from inside the `tests-dev` directory. Currently, both cases are enabled on Orion and Hera, but will be ported to other Tier 1 platforms in the near term.

Configuration details:

2020 July CAPE:

- ATM-only
- Resolution: C48
- default CCPP suite: `FV3_GFS_v16`
- ICs: GFSv16 input data generated manually with UFS_UTILS (v2 sfc files)

Baroclinic wave:
- ATM-only
- Resolution: C192
- default CCPP suite: (dycore only)
- ICs: currently uses default for C192 from `INPUT_L127_gfsv17`; `INPUT` data is only used to enable model to run

The baroclinic wave case utilizes a model configuration provided by @XiaqiongZhou-NOAA at NOAA EMC, who also kindly provided information and an example of how to configure the case. The input namelist for this case can be found [here](https://github.com/ulmononian/ufs-weather-model/blob/feature/enable_test_cases/tests-dev/test_cases/parm/baroclinic_wave.nml.IN).

Testing and porting of this work done in collaboration with @natalie-perlin.

### Commit Message:
```
* UFSWM - Add tests-dev ATM-only idealized dry baroclinic wave test and a 2020 July CAPE case
* UFSWM - Update modulefile to support Rocky 8 on CSPs, with ParallelWorks

```
### Priority:
* Normal

### UFSWM:

* Closes #2309 

### Sub component Pull Requests:
* None

### UFSWM Blocking Dependencies:
* None

---
### Regression Test Changes (Please commit test_changes.list):
* PR Adds New Tests/Baselines.

### Input data Changes:

* New input data: GFS ICs for 2020 July CAPE case; `global_hyblev.l128.txt` file for baroclinic wave remapping

### Library Changes/Upgrades:
* No Updates
  
---

## Testing Log:
- RDHPCS
  - [x] Hera
  - [x] Orion
  - [x] Hercules
  - [x] Jet
  - [ ] Gaea
  - [ ] Derecho
- WCOSS2
  - [ ] Dogwood/Cactus
  - [ ] Acorn
- [ ] CI
- [ ] opnReqTest (complete task if unnecessary)

### CODE REVIEW

**Comment by ulmononian** [comment]
Full Hera RT suite (Intel/GNU) 
[RegressionTests_hera.log](https://github.com/user-attachments/files/17369327/RegressionTests_hera.1.log). Includes two new idealized cases.

File: tests-dev/test_cases/test_cases.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+test_cases_intel:
```
Comment (jkbk2004) [inline tests-dev/test_cases/test_cases.yaml]: why do we need separate test case application, if https://github.com/ufs-community/ufs-weather-model/blob/develop/tests-dev/ufs_test.yaml#L108 is same build configuration ?

File: tests-dev/test_cases/exp_conf/2020_CAPE.IN [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  mkdir INPUT RESTART
+  if [ "$NESTED" = .true. ]; then
+    cp -r @[INPUTDATA_ROOT]/${inputdir}/INPUT_L127_nested/* ./INPUT/.
+  elif [ "$V2_SFC_FILE" = "true" ]; then
+    cp -r @[INPUTDATA_ROOT]/${inputdir}/INPUT_L127_v2_sfc/* ./INPUT/.
+  elif [ "$HSD_CASE" = "true" ]; then
+    if [ ${MACHINE_ID} = hera ]; then
+      cp -r /scratch1/NCEPDEV/nems/role.epic/HSD_INPUT_DATA/FV3_input_data48/INPUT_L127_v2_sfc/* ./INPUT/.
+    elif [ ${MACHINE_ID} = orion ]; then
+     cp -r /work/noaa/epic/role-epic/contrib/HSD_INPUT_DATA/FV3_input_data48/INPUT_L127_v2_sfc/* ./INPUT/.
```
Comment (jkbk2004) [inline tests-dev/test_cases/exp_conf/2020_CAPE.IN]: We don't recommend to maintain multiple hard-wired input file paths, especially machine specifically. New input file requirement should be specified in the PR description. We can help to add something like INPUTDATA_HSD in current baseline directories across machines.

Comment (jkbk2004) [reply tests-dev/test_cases/exp_conf/2020_CAPE.IN]: It looks like we have identical file sets- at /scratch2/NAGAPE/epic/UFS-WM_RT/NEMSfv3gfs/input-data-20240501/FV3_input_data48/INPUT_L127_v2_sfc. What is difference at /scratch1/NCEPDEV/nems/role.epic/HSD_INPUT_DATA/FV3_input_data48/INPUT_L127_v2_sfc/* ?

Comment (jkbk2004) [reply tests-dev/test_cases/exp_conf/2020_CAPE.IN]: If identical, should the new cases this pr is adding need to come along with https://github.com/ufs-community/ufs-weather-model/blob/develop/tests-dev/ufs_test.yaml#L119 ?

**Comment by jkbk2004** [comment]
@WenMeng-NOAA #2489 is combined to this pr.

**Comment by WenMeng-NOAA** [comment]
> @WenMeng-NOAA #2489 is combined to this pr.

@jkbk2004 Thanks!

**Comment by WenMeng-NOAA** [comment]
> > @WenMeng-NOAA #2489 is combined to this pr.
> 
> @jkbk2004 Thanks!

@ulmononian Could you incorporate [this change](https://github.com/ufs-community/ufs-weather-model/blob/803c43d115eb3330ab00547ff69c33651b62fe94/tests/tests/rap_clm_lake_debug#L26) from #2489? Thanks!

**Comment by jkbk2004** [comment]
Rocoto is not stable on Gaea and Derecho. But baselines of develop-20241031 on those machines are ok.

File: tests-dev/create_log.py [hunk 3]
```diff
@@ -3,10 +3,27 @@
 import subprocess
 import yaml
 from datetime import datetime
+#import datetime
 from ufs_test_utils import get_testcase, write_logfile, delete_files, machine_check_off
 
+def get_timestamps(path):
+    """Obtain experiment starting and ending time marks through file timestapns
```
Comment (NickSzapiro-NOAA) [inline tests-dev/create_log.py]: timestapns --> timestamps

File: tests-dev/test_cases/exp_conf/2020_CAPE.IN [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+    RFILE=ufs.atmw.ww3.r.${RESTART_FILE_SUFFIX_SECS}
+    cp ../${DEP_RUN}${SUFFIX}/${RFILE} .
+  fi
+
+  if [ $MODEL_INITIALIZATION = "false" ]; then
+    #read_inc and iau tests restart from fh=24
+    CURRDIR=`pwd`
+    FILEDIR=${CURRDIR}/../${DEP_RUN}${SUFFIX}/RESTART
+  else
+    #for warm start from initialiation time
```
Comment (NickSzapiro-NOAA) [inline tests-dev/test_cases/exp_conf/2020_CAPE.IN]: initialiation --> initialization

File: tests-dev/test_cases/exp_conf/2020_CAPE.IN [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+      # CMEPS restart and pointer files
+      RFILE=ufs.atmw.cpl.r.${RESTART_FILE_SUFFIX_SECS}.nc
+      cp  ../${DEP_RUN}${SUFFIX}/RESTART/${RFILE} .
+      ls -1 ${RFILE}>rpointer.cpl
+    fi
+    RFILE=ufs.atmw.ww3.r.${RESTART_FILE_SUFFIX_SECS}
+    cp ../${DEP_RUN}${SUFFIX}/${RFILE} .
+  fi
+
+  if [ $MODEL_INITIALIZATION = "false" ]; then
```
Comment (NickSzapiro-NOAA) [inline tests-dev/test_cases/exp_conf/2020_CAPE.IN:50]: I don't see why MODEL_INITIALIZATION is needed. Isn't this just if it is a DEP_RUN?

File: tests-dev/test_cases/exp_conf/baroclinic_wave.IN
Comment (NickSzapiro-NOAA) [inline tests-dev/test_cases/exp_conf/baroclinic_wave.IN:1]: If exp_conf/2020_CAPE.IN and exp_conf/baroclinic_wave.IN are not duplicates, they look very similar. Maybe consider combining them or building off base configurations sequentially

Comment (ulmononian) [reply tests-dev/test_cases/exp_conf/baroclinic_wave.IN:1]: @NickSzapiro-NOAA thanks for your review! we will look into consolidating and optimizing the *IN files for these cases in follow-up work.

File: tests-dev/test_cases/exp_conf/2020_CAPE.IN (hunk 4 above)
Comment (ulmononian) [reply tests-dev/test_cases/exp_conf/2020_CAPE.IN]: fixed

File: tests-dev/create_log.py (hunk 3 above)
Comment (ulmononian) [reply tests-dev/create_log.py]: fixed

File: tests-dev/test_cases/exp_conf/2020_CAPE.IN (hunk 5 above)
Comment (ulmononian) [reply tests-dev/test_cases/exp_conf/2020_CAPE.IN:50]: i can look into trimming this down in a follow-up

**Comment by ulmononian** [comment]
@NickSzapiro-NOAA would you be able to resolve the comments if you are ok with the replies?

File: tests-dev/test_cases/exp_conf/2020_CAPE.IN (hunk 5 above)
Comment (jkbk2004) [reply tests-dev/test_cases/exp_conf/2020_CAPE.IN:50]: sounds like new PR will come soon to update documentation and container.

**Review by jkbk2004 — APPROVED** [review]

**Review by zach1221 — APPROVED** [review]

**Review by FernandoAndrade-NOAA — APPROVED** [review]

**Comment by NickSzapiro-NOAA** [comment]
@ulmononian I don't have permissions to resolve comments. Anyways, thanks for the revisions!

**Pull request merged**